daemonset-updater/main.go

```diff
@@ -0,0 +1,283 @@
+		return
+	}
+
+	// get node lists per a rack
+	rackList := make(map[int][]string)
+	for _, node := range nodeList.Items {
+		// ignore the node whether it is in the ignore list
+		if slices.Contains(cfg.Ignore, node.GetName()) {
+			log.InfoContext(ctx, "ignore to update", slog.String("node", node.GetName()))
+			continue
+		}
+		rackNumber, ok := node.Labels[RackLabel]
+		if !ok {
+			log.WarnContext(ctx, "the rack label is not found", slog.String("node", node.GetName()))
+			continue
+		}
+		n, err := strconv.Atoi(rackNumber)
+		if err != nil {
+			log.WarnContext(ctx, "the value of the rack label is invalid", slog.String("node", node.GetName()), slog.String("value", rackNumber))
+			continue
+		}
+		if _, ok = rackList[n]; !ok {
+			rackList[n] = []string{node.GetName()}
+		} else {
+			rackList[n] = append(rackList[n], node.GetName())
+		}
+	}
+
+	// get a pod list of the daemonset
+	podList := corev1.PodList{}
+	listOption := client.MatchingLabels{
+		appLabelKey: appLabelValue,
+	}
+	if err := c.List(ctx, &podList, listOption); err != nil {
+		log.ErrorContext(ctx, "failed to list pods", slog.Any("error", err))
+		return
+	}
+
+	podNodeMap := make(map[string]*corev1.Pod)
+	for _, pod := range podList.Items {
+		podNodeMap[pod.Spec.NodeName] = &pod
+	}
+
+	// get nodes that is not desired image per a rack
+	updateList := make(map[int][]string)
+	for rackNumber, nodeList := range rackList {
+		updateNodeList := make([]string, 0, len(nodeList))
+		for _, node := range nodeList {
+			pod, ok := podNodeMap[node]
+			if !ok {
+				log.WarnContext(ctx, "failed to find the corresponding pod to the node", slog.String("node", node))
+				continue
+			}
+
+			// check the image is the desired one.
+			// should we use .spec.containers[0]?
+			image := pod.Spec.Containers[0].Image
+			if image != cfg.DesiredImage {
+				log.InfoContext(ctx, "image is not desired", slog.String("node", node), slog.String("pod", pod.GetName()))
+				updateNodeList = append(updateNodeList, node)
+			}
+		}
+		if len(updateNodeList) > 0 {
+			updateList[rackNumber] = updateNodeList
+		}
+		log.InfoContext(ctx, "the rack has nodes that is needed to update", slog.Int("rack", rackNumber), slog.Any("nodes", updateNodeList))
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	// var wg sync.WaitGroup
+
+	targetChan := make(chan string)
+	complete := make(chan struct{})
+	go func() {
+		for rackNumber, nodeList := range updateList {
+			log.InfoContext(ctx, "start to update", slog.Int("rack", rackNumber))
+
+			for _, node := range nodeList {
+				targetChan <- node
+				// drain(dry-run)
+				log.InfoContext(ctx, "start to drain", slog.String("node", node), slog.Bool("dry-run", true))
+				if err := drain(ctx, node, cfg.DrainTimeout, true); err != nil {
+					log.WarnContext(ctx, "failed to drain", slog.String("node", node), slog.Bool("dry-run", true), slog.Any("error", err))
+					// if we fail to drain, even if it is dry-run, go next.
+					continue
+				}
```

Comment:
When the context is canceled, the iteration doesn't seem to end until the last node is reached.
(the impact looks minor as it just produces a large amount of "failed to drain" logs.)